File: leetcode/0389-find-the-difference/findTheDifference.ts
function findTheDifference(s: string, t: string): string {
  const letterCountMap = new Map<string, number>();
  let result = "";

  for (const string of s) {
    if (letterCountMap.has(string)) {
      letterCountMap.set(string, letterCountMap.get(string)! + 1);
      continue;
    }

    letterCountMap.set(string, 1);
  }

  for (const string of t) {
    if (!letterCountMap.has(string)) {
      result = string;
      break;
    }

    const letterCount = letterCountMap.get(string)!;
    if (letterCount < 1) {
      result = string;
      break;
    }

    letterCountMap.set(string, letterCount - 1);
  }

  return result;
}

export default findTheDifference;

File: leetcode/0389-find-the-difference/findTheDifference.go
package leetcode

func findTheDifference(s string, t string) byte {
	var result byte
	var letterCountMap = map[rune]int{}

	for _, letter := range s {
		if count, ok := letterCountMap[letter]; ok {
			letterCountMap[letter] = count + 1
			continue
		}

		letterCountMap[letter] = 1
	}

	for _, letter := range t {
		var count, ok = letterCountMap[letter]

		if !ok || count <= 0 {
			result = byte(letter)
			break
		}

		letterCountMap[letter] = count - 1
	}

	return result
}
